nispor: Introduce nispor plugin

cathay4t opened the pull request:
Introduce `NisporPlugin` for querying on these informations using
python3-nispor package:
 * MTU, MAC, state, IP address.
 * DHCP/AUTOCONF status:
    * Mark DHCP/AUTOCONF as true when found IP address with valid_lft not
      forever and correct prefix length.

As nispor might be wrong on DHCP status and interface type comparing to
NetworkManager, so NetworkManager plugin take priority over nispor
plugin and override conflict data.

Removed the code in NetworkManager plugin for querying above information
except the DHCP/Autoconf and IP stack enable status.

Integration test case added.

Signed-off-by: Gris Ge <fge@redhat.com>

ffmancera commented on libnmstate/nispor/base_iface.py:
```diff
@@ -0,0 +1,140 @@
+
+    @property
+    def mac(self):
+        return self._np_info.get("mac_address", "00:00:00:00:00:00")
```
We are using hardcoded strings to access to the different values. Maybe it would be nice to have an schema API from Nispor. I know this request is not part of this PR but I am just asking around :-)

cathay4t replied:
I think the better approach would be nispor python module providing objects.
We should use property of class instead of a dictionary.

Will track this in:
https://github.com/nispor/nispor/issues/10

ffmancera replied:
Nice! Thanks!

cathay4t replied:
Changed to using object property.

ffmancera commented on libnmstate/nm/applier.py:
```diff
@@ -81,6 +81,7 @@ def apply_changes(context, net_state, save_to_disk):
 
     _preapply_dns_fix(context, net_state)
     _mark_nm_external_subordinate_changed(context, net_state)
+    _mark_nm_unmanaged_ubordinate_changed(context, net_state)
```
There is a typo: "_mark_nm_unmanaged\_**s**ubordinate_changed"

ffmancera commented on libnmstate/nm/applier.py:
```diff
@@ -655,3 +656,13 @@ def _mark_nm_external_subordinate_changed(context, net_state):
                         subordinate_iface.mark_as_changed()
+
+
+def _mark_nm_unmanaged_ubordinate_changed(context, net_state):
```
Same typo here.

ffmancera replied:
Is this needed for this PR? I do not understand how is this related to the nispor support. If it is related, please add the explanation.

cathay4t replied:
It will fail `test_linux_bridge_replace_unmanaged_port`, let me try to extract it to another pull request.

cathay4t replied:
The root cause of that is https://bugzilla.redhat.com/show_bug.cgi?id=1869063

Before using nispor, any unmanaged interface is showing as DOWN state with both ipv6/ipv4 disabled.
After using nispor, the unmanaged interface will show its real  link state and ipv4/ipv6 info.
In that test case, the unmanaged dummy0 is up with ipv6 link-local. Due to above NetworkManager bug, NetworkManager will refused to activated a ipv6 link-local method to dummy0 as it was a bridge port.

cathay4t replied:
Also impacted by https://bugzilla.redhat.com/show_bug.cgi?id=1869079

cathay4t replied:
For this PR. The original workaround code has been removed.

ffmancera replied:
Nice, thanks!

ffmancera commented on libnmstate/nmstate.py:
```diff
@@ -99,7 +100,7 @@ def plugins_capabilities(plugins):
 
 def _load_plugins():
-    plugins = [NetworkManagerPlugin()]
+    plugins = [NetworkManagerPlugin(), NisporPlugin()]
```
Is the order important here?

cathay4t replied:
No. The plugin has its priority.

ffmancera replied:
Thanks!

ffmancera commented on tests/integration/nm/ipv4_test.py:
```diff
@@ -40,10 +40,19 @@
 
 
+def _dict_is_subset(superset, subset):
+    """
+    Code copied from Ignacio Vazquez-Abrams in
+    https://stackoverflow.com/questions/9323749
+    License is CC BY-SA 3.0
+    """
+    return all(item in superset.items() for item in subset.items())
```
Not sure if this is needed. Or if that piece of code could be copyrighted. In addition, we need to check if CC license is compatible with LGPL-2.1.

ffmancera replied:
I checked out and it is compatible.

cathay4t replied:
I checked also. I put it there is to honor the original developer.

ffmancera replied:
Sure, I am fine with it.

ffmancera commented on tests/integration/nm/linux_bridge_test.py:
```diff
@@ -268,26 +267,8 @@ def _remove_read_only_properties(bridge_state):
-            ]
-        },
-        verify_change=False,
-    )
+    with nm_unmanaged_dummy(DUMMY1):
+        yield
```
Can we use directly `nm_unmanaged_dummy`?

cathay4t replied:
The I need to do the `try-finally` stuff which is done by pytest fixture.

ffmancera replied:
Oh I see. I am fine with this then.

ffmancera commented on tests/integration/nmstatectl_test.py:
```diff
@@ -97,6 +113,17 @@
 
+@pytest.fixture(scope="module", autouse=True)
+def enable_lo_ipv6():
+    with open("/proc/sys/net/ipv6/conf/lo/disable_ipv6") as fd:
+        old_state = fd.read().strip()
+    with open("/proc/sys/net/ipv6/conf/lo/disable_ipv6", "w") as fd:
+        fd.write("0")
```
I think you can do it just opening the file one time.

cathay4t replied:
Yes. Changed to use `r+` mode.

cathay4t commented:
Patch ready for review.

cathay4t commented:
Wait PR #1278

cathay4t committed:
nispor: Introduce nispor plugin

Introduce `NisporPlugin` for querying on these informations using
python3-nispor package:
 * MTU, MAC, state, IP address.
 * DHCP/AUTOCONF status:
    * Mark DHCP/AUTOCONF as true when found IP address with valid_lft not
      forever and correct prefix length.

As nispor might be wrong on DHCP status and interface type comparing to
NetworkManager, so NetworkManager plugin take priority over nispor
plugin and override conflict data.

Removed the code in NetworkManager plugin for querying above information
except the DHCP/Autoconf and IP stack enable status.

Integration test case added.

Signed-off-by: Gris Ge <fge@redhat.com>

ffmancera reviewed (approved):
LGTM, thanks!
